src/backend/pipeline/cont_adhoc.c

```diff
@@ -0,0 +1,1000 @@
+					 	 Tuplestorestate *batch)
+{
+	TuplestoreScan *scan;
+
+	Relation rel = heap_openrv(matrel, AccessShareLock);
+
+	plan->is_continuous = false;
+
+	scan = SetCombinerPlanTuplestorestate(plan, batch);
+	scan->desc = CreateTupleDescCopy(RelationGetDescr(rel));
+
+	heap_close(rel, AccessShareLock);
+
+	return scan->desc;
+
+}
+
+/* Get the combiner plan, and prepare the combiner state for execution */
+static AdhocCombinerState *
+init_adhoc_combiner(ContinuousViewData data,
+					Tuplestorestate *batch)
+{
+	AdhocCombinerState *state = palloc0(sizeof(AdhocWorkerState));
+	PlannedStmt *pstmt = 0;
+
+	PushActiveSnapshot(GetTransactionSnapshot());
+	pstmt = GetContPlan(data.view, Combiner);
+	PopActiveSnapshot();
+
+	state->tup_desc = prepare_plan_for_reading(data.view->matrel, pstmt, batch);
+	state->dest = CreateDestReceiver(DestTuplestore);
+	 
+	state->query_desc = CreateQueryDesc(pstmt, NULL, InvalidSnapshot,
+										InvalidSnapshot, state->dest,
+										NULL, 0);
+
+	state->slot = MakeSingleTupleTableSlot(state->tup_desc);
+
+	state->batch = batch;
+
+	state->result = tuplestore_begin_heap(true, true,
+							continuous_query_combiner_work_mem);
+
+	SetTuplestoreDestReceiverParams(state->dest, state->result,
+									CurrentMemoryContext, true);
+
+	/* if this is an aggregating query, set up a hash table to store the
+	 * groups */
+
+	if (IsA(pstmt->planTree, Agg))
+	{
+		Agg *agg = (Agg *) pstmt->planTree;
+		state->is_agg = true;
+
+		state->existing = create_agg_hash(agg->numCols, agg->grpOperators,
+										  agg->grpColIdx);
+	}
+
+	ExecutorStart(state->query_desc, 0);
+	state->query_desc->estate->es_lastoid = InvalidOid;
+
+	(*state->dest->rStartup) (state->dest, 
+			state->query_desc->operation, state->query_desc->tupDesc);
+
+	ExecEndNode(state->query_desc->planstate);
+	FreeExecutorState(state->query_desc->estate);
+
+	return state;
+}
+
+/* run the worker - this will block for one sec waiting for tuples in the worker tuple
+ * buffer. 
+ *
+ * If it has no tuples, it will return false.
+ *
+ * Otherwise, tuples are processed with the worker plan and sent to the dest receiver
+ */
+static bool
+exec_adhoc_worker(AdhocWorkerState * state)
+{
+	struct Plan *plan = 0;
+	bool has_tup = false;
+	EState *estate = 0;
+	ResourceOwner owner = CurrentResourceOwner;
+
+	TupleBufferBatchReaderTrySleepTimeout(state->reader, state->last_processed, 1000);
+
+	has_tup = 
+		TupleBufferBatchReaderHasTuplesForCQId(state->reader, state->view_id);
+
+	if (!has_tup)
+	{
+		TupleBufferBatchReaderRewind(state->reader);
+		return false;
+	}
+
+	state->query_desc->estate = create_estate(state->query_desc);
+	estate = state->query_desc->estate;
+
+	TupleBufferBatchReaderSetCQId(state->reader, state->view_id);
+
+	set_snapshot(estate, owner);
+
+	plan = state->query_desc->plannedstmt->planTree;
+
+	state->query_desc->planstate = 
+		ExecInitNode(plan, state->query_desc->estate, 0);
+
+	set_reader(state->query_desc->planstate, state->reader);
```

Comment:
Do you mean the call to set_reader? Because the planstate is being reinitialized by ExecInitNode on each run, so set_reader does need to be called.